### PR TITLE
Fix and document BIO_FLAGS_NONCLEAR_RST behavior on memory BIO

### PR DESCRIPTION
**Pull request by t8m**
The BIO_FLAGS_NONCLEAR_RST flag behavior was not properly documented
and it also caused the length to be incorrectly set after the reset
operation.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: doc/man3/BIO_s_mem.pod [hunk 1]
```diff
@@ -104,6 +105,12 @@ BIO is set to BIO_NOCLOSE, before freeing the BUF_MEM the data pointer
 in it must be set to NULL as the data pointer does not point to an
 allocated memory.
 
+Calling BIO_reset() on a read write memory BIO with BIO_FLAGS_NONCLEAR_RST
+flag set can have unexpected outcome when the reads and writes to the
+BIO are intertwined. As documented above the BIO will be reset to the
+state after the last completed write operation. The effects of reads
+preceeding that write operation cannot be undone.
```
Comment (bernd-edlinger) [inline doc/man3/BIO_s_mem.pod]: I assume that BIO_get_mem_ptr changes the reset state, right?

Comment (t8m) [reply doc/man3/BIO_s_mem.pod]: Good point, I've added a paragraph documenting it.

File: doc/man3/BIO_s_mem.pod [hunk 2]
```diff
@@ -104,6 +105,15 @@ BIO is set to BIO_NOCLOSE, before freeing the BUF_MEM the data pointer
...
+
+Calling BIO_get_mem_ptr() prior to a BIO_reset() call with
+BIO_FLAGS_NONCLEAR_RST set has the same effect as a write operation.
```
Comment (bernd-edlinger) [inline doc/man3/BIO_s_mem.pod]: should we note that BIO_get_mem_data does not have the same effect?

Comment (t8m) [reply doc/man3/BIO_s_mem.pod]: I do not see it as necessary. It is implied by not mentioning it here. But if you think it is necessary to mention it explicitly please tell me and I can add it.

Comment (bernd-edlinger) [reply doc/man3/BIO_s_mem.pod]: ok, fine.

**Comment by t8m** [comment]
@openssl/omc Please review

**Comment by t8m** [comment]
[master 8b7b32921] Fix and document BIO_FLAGS_NONCLEAR_RST behavior on memory BIO
[OpenSSL_1_1_1-stable c6991655c4] Fix and document BIO_FLAGS_NONCLEAR_RST behavior on memory BIO

**Pull request closed without merging**